chemicalx/models/gcnbmp.py

```diff
@@ -1,17 +1,263 @@
+        g = scatter_add(g, graph_index, dim=0)
+        return g
+
+
+class GCNBMPEncoder(nn.Module, core.Configurable):
+    """
+    The drug encoding backbone from [gcnbmp].
+
+    .. [gcnbmp] `GCN-BMP: Investigating graph representation learning for DDI prediction task
+       <https://www.sciencedirect.com/science/article/pii/S1046202320300608>`_
+
+    """
+
+    def __init__(
+        self,
+        input_dim: int,
+        hidden_dims: List[int],
+        num_relation: int,
+        edge_input_dim: Optional[int] = None,
+        batch_norm: Optional[bool] = False,
+        activation: Optional[str] = "sigmoid",
+    ):
+        """
+        Instantiate the GCN-BMP encoder.
+
+        :param input_dim: input dimension.
+        :param hidden_dims: hidden dimensions.
+        :param num_relation: number of relations.
+        :param edge_input_dim: dimension of edge features.
+        :param batch_norm: apply batch normalization on nodes or not.
+        :param activation: activation function.
+        """
+        super(GCNBMPEncoder, self).__init__()
+
+        if not isinstance(hidden_dims, Sequence):
+            hidden_dims = [hidden_dims]
+        self.input_dim = input_dim
+        self.dims = [input_dim] + list(hidden_dims)
+        self.num_relation = num_relation
+
+        self.layers = nn.ModuleList()
+        for left_dim, right_dim in pairwise(self.dims):
+            self.layers.append(
+                layers.RelationalGraphConv(left_dim, right_dim, num_relation, edge_input_dim, batch_norm, activation)
+            )
+            self.layers.append(Highway(right_dim, left_dim))
+
+    def forward(self, graph: torchdrug.data.graph.PackedGraph, input: torch.Tensor) -> dict:
+        """
+        Compute the node representations and the graph representation(s).
+
+        Require the graph(s) to have the same number of relations as this module.
+
+        Parameters:
+            graph: :math:`n` graph(s)
+            input: input node representations
+
+        Returns:
+            dict with ``node_feature`` field:
+                node representations of shape :math:`(|V|, d)`
+        """
+        hiddens = []
+        layer_input = input
+        prev_gcn = input
+
+        for conv, highway in chunked(self.layers, 2):
+            hidden = conv(graph, layer_input)
+            hiddens.append(hidden)
+            hidden = highway(hidden, prev_gcn)
+            hiddens.append(hidden)
+            layer_input = hidden
+            prev_gcn = hiddens[-2]
+
+        node_feature = hiddens[-1]
+        return {"node_feature": node_feature}
 
-class GCNBMP(UnimplementedModel):
-    """An implementation of the GCN-BMP model from [chen2020]_.
 
-    .. seealso:: This model was suggested in https://github.com/AstraZeneca/chemicalx/issues/21
+class GCNBMP(Model):
+    r"""The GCNBMP model from [gcnbmp].
 
-    .. [chen2020] Chen, X., *et al.* (2020). `GCN-BMP: Investigating graph representation learning
```

Comment:
please be careful when merging - the citation on the main branch is the desired one.